src/tools/miri/tests/pass-dep/concurrency/linux-futex.rs

```diff
@@ -284,6 +284,78 @@ fn concurrent_wait_wake() {
     assert!(woken > 0 && woken < rounds);
 }
 
+// Crucial test which relies on the SeqCst fences in futex lock/unlock.
+fn concurrent_lock_unlock() {
+    const fn unlocked() -> u32 {
+        0
+    }
+
+    fn locked() -> u32 {
+        unsafe { libc::gettid() }.try_into().unwrap()
+    }
+
+    static FUTEX: AtomicU32 = AtomicU32::new(0);
+    static mut DATA: u32 = 0;
+    static WOKEN: AtomicU32 = AtomicU32::new(0);
+
+    let rounds = 50;
+    for _ in 0..rounds {
+        unsafe { DATA = 0 }; // Reset
+        // Suppose the main thread is holding a lock implemented using futex...
+        FUTEX.store(locked(), Ordering::Relaxed);
+
+        let t = thread::spawn(move || {
+            unsafe {
+                // FUTEX_LOCK_PI guarantees we do not wake before the main thread unlocks.
+                let ret = libc::syscall(
+                    libc::SYS_futex,
+                    addr_of!(FUTEX),
+                    libc::FUTEX_LOCK_PI,
+                    0, // unused
+                    ptr::null::<libc::timespec>(),
+                );
+                // `ret` is always 0 on success.
+                assert_eq!(ret, 0);
+                // Check that we own the lock.
+                assert_eq!(FUTEX.load(Ordering::Relaxed), locked());
+                // We actually slept. And then woke up again. So we should be ordered-after
+                // what happened-before the FUTEX_WAKE. So this is not a race.
+                assert_eq!(DATA, 1);
+                // Also remember that this happened at least once.
+                WOKEN.fetch_add(1, Ordering::Relaxed);
```

Comment:
This is incremented *each time*, right? That's odd. Why will `FUTEX_LOCK_PI` always succeed?

In the original test, the syscall can fail. So it seems you fundamentally changed the nature of the test here?

Reply:
Yes, `FUTEX_LOCK_PI` always succeeds.

For `FUTEX_WAIT` in the original test:
- If it is contended, `FUTEX == HELD`, syscall put the waiter thread into sleep.
- If it is not, `FUTEX == FREE != HELD`, syscall fails immediately.

For `FUTEX_LOCK_PI` in this test:
- If it is contended, syscall changes `FUTEX` from the owner thread's `tid` to `tid | FUTEX_WAITERS` (this is test in the `loop` below), and put the waiter thread into sleep.
- If it is not, `FUTEX == 0`, syscall changes `FUTEX` from `0` to the new owner's `tid`, and returns immediately.

So all state changes are done in the kernel. `FUTEX_LOCK_PI` always succeeds.

Reply:
> So it seems you fundamentally changed the nature of the test here?

Fixed in the latest commit. I added the test of memory ordering.

Reply:
>  put the waiter thread into sleep

Yeah that's the point, the test should sometimes block ("succeed" is what I said above because this happens when the comparison is "successful", i.e., returns true) word) and sometimes not block ("fail").

Reply:
Please keep this unresolved until I am happy with the testcase. it's so hard to find your messages otherwise as github hides resolved comments.